quinn-h3/src/client.rs

```diff
@@ -94,3 +107,97 @@ impl Future for Connecting {
         )))
     }
 }
+
+enum SendRequestState {
+    Opening(OpenBi),
+    Sending(WriteAll<SendStream, Vec<u8>>),
+    Sent(Shutdown<SendStream>),
+    Receiving(FrameStream<RecvStream>),
+    Ready(HeadersFrame),
+    Finished,
+}
+
+pub struct SendRequest {
+    req: Request<()>,
+    state: SendRequestState,
+    conn: ConnectionRef,
+    recv: Option<FrameStream<RecvStream>>,
+}
+
+impl SendRequest {
+    fn new(req: Request<()>, open_bi: OpenBi, conn: ConnectionRef) -> Self {
+        Self {
+            req,
+            conn,
+            state: SendRequestState::Opening(open_bi),
+            recv: None,
+        }
+    }
+}
+
+impl Future for SendRequest {
+    type Item = RecvResponse;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            match self.state {
+                SendRequestState::Opening(ref mut o) => {
+                    let (send, recv) = try_ready!(o.poll());
+                    self.recv = Some(FrameStream::new(recv));
+
+                    let header = {
+                        let mut conn = self.conn.inner.lock().unwrap();
+                        conn.encode_header(&send.id(), self.req.headers())?
+                    };
+                    let mut encoded_header = vec![];
+                    header.encode(&mut encoded_header);
+
+                    let send = io::write_all(send, encoded_header);
+                    self.state = SendRequestState::Sending(send);
+                }
+                SendRequestState::Sending(ref mut send) => {
+                    let (send, _) = try_ready!(send.poll());
+                    let shut = io::shutdown(send);
+                    self.state = SendRequestState::Sent(shut);
+                }
+                SendRequestState::Sent(ref mut shut) => {
+                    try_ready!(shut.poll());
+                    self.state = match mem::replace(&mut self.recv, None) {
+                        Some(r) => SendRequestState::Receiving(r),
+                        None => return Err(Error::Internal("Invalid receive state")),
+                    }
+                }
+                SendRequestState::Receiving(ref mut frames) => match try_ready!(frames.poll()) {
+                    None => return Err(Error::peer("recieved an empty response")),
```

Comment:
`recieved` -> `received` (I think you have this typo in multiple places, also in commit messages).

Reply:
Fixed in 60b761c